### PR TITLE
Join/leave flow with majority height

### PR DESCRIPTION
**Pull request by marcinwadon**

### CODE REVIEW

File: src/main/scala/org/constellation/domain/redownload/RedownloadService.scala [hunk 1]
```diff
@@ -520,6 +549,9 @@ object RedownloadService {
   type SnapshotsAtHeight = Map[Long, String] // height -> hash
   type SnapshotProposalsAtHeight = Map[Long, SnapshotProposal]
   type PeersProposals = Map[Id, SnapshotProposalsAtHeight]
+  type PeersCache = Map[Id, MajorityHeight]
   type SnapshotInfoSerialized = Array[Byte]
   type SnapshotSerialized = Array[Byte]
+
+  case class LatestMajorityHeight(low: Long, high: Long)
```
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/redownload/RedownloadService.scala]: `lowest`, `highest` or `min` `max` imo.

`low` or `high` doesn't mean `lowest` and `highest`

File: src/main/scala/org/constellation/util/CommonEndpoints.scala [hunk 2]
```diff
@@ -164,6 +165,12 @@ trait CommonEndpoints extends Json4sSupport {
       path("state") {
         APIDirective.handle(dao.cluster.getNodeState)(res => complete(NodeStateInfo(res, dao.addresses, dao.nodeType)))
       } ~
+      path("latestMajorityHeight") {
+        val height = (dao.redownloadService.lowestMajorityHeight, dao.redownloadService.latestMajorityHeight)
+          .mapN(LatestMajorityHeight)
```
Comment (kpudlik) [inline src/main/scala/org/constellation/util/CommonEndpoints.scala]:
```scala
val height = LatestMajorityHeight(
  dao.redownloadService.lowestMajorityHeight,
  dao.redownloadService.latestMajorityHeight,
)
```

Imo it is kinda overkill ;p

File: src/main/scala/org/constellation/domain/redownload/RedownloadService.scala [hunk 3]
```diff
@@ -437,10 +460,12 @@ class RedownloadService[F[_]: NonEmptyParallel](
     majorityState: SnapshotsAtHeight,
     redownloadInterval: Int
   ): Boolean =
-    getAlignmentResult(acceptedSnapshots, majorityState, redownloadInterval) match {
-      case AlignedWithMajority => false
-      case _                   => true
-    }
+    if (majorityState.isEmpty) false
```
Comment (kpudlik) [inline src/main/scala/org/constellation/domain/redownload/RedownloadService.scala]: I was thinking about new alignment result like `EmptyMajority`.
We would then have:

```scala
getAlignmentResult(acceptedSnapshots, majorityState, redownloadInterval) match {
        case AlignedWithMajority => false
        case EmptyMajority  => false
        case _  => true
}
```

But maybe explicit check before makes more sense 👍

File: src/main/scala/org/constellation/domain/redownload/RedownloadService.scala (hunk 1 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/domain/redownload/RedownloadService.scala]: done

File: src/main/scala/org/constellation/util/CommonEndpoints.scala (hunk 2 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/util/CommonEndpoints.scala]: It's `F` :)

File: src/main/scala/org/constellation/domain/redownload/RedownloadService.scala (hunk 3 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/domain/redownload/RedownloadService.scala]: :+1:

**Pull request merged**